### PR TITLE
Ruff violation icn001 fix

### PR DESCRIPTION
**Pull request by Shaheer-Ahmd**
### Description

This PR fixes the violation of RUFF rule ICN001
# unconventional-import-alias (ICN001)

Derived from the **flake8-import-conventions** linter.

Fix is sometimes available.

## What it does
Checks for imports that are typically imported using a common convention,
like `import pandas as pd`, and enforces that convention.

## Why is this bad?
Consistency is good. Use a common convention for imports to make your code
more readable and idiomatic.

For example, `import pandas as pd` is a common
convention for importing the `pandas` library, and users typically expect
Pandas to be aliased as `pd`.

## Example
```python
import pandas
```

Use instead:
```python
import pandas as pd
```

This pull request is to address ...

Fixes #14818

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by nstarman** [comment]
Hi @Shaheer-Ahmd. Thanks for opening the PR! I like the rule ICN001 and think it's worth implementing. This PR is quite ambitious in that it fixes ICN001 everywhere in Astropy. In [Ensure compliance with more Ruff rules](https://github.com/astropy/astropy/issues/14818) we actually recommend breaking these large PRs into a few smaller ones. That way not everyone in Astropy has to sign off on an omnibus PR, instead you can get these PRs in faster by needing only 1 or 2 people to approve for each PR.

**Comment by dhomeier** [comment]
Agreed that it might be difficult to get approval from 10 reviewers, though this could be tricky to break into sensible-sized chunks, as many subpackages have exactly 1 or 2 violations. Noting two parts that could perhaps be treated separately:
1. A very large part of the `numpy` and `matplotlib` violations seem to sit in test config and `setup_package` files – no idea if that was a convention at some point.
2. `units` stands out with all its `_np` or `_numpy` imports, most of which date back at least to #370. I could find any reason for that in the discussion, but possibly the point was to treat all numpy constants as private. Don't really see how that might make sense though, as `np` is available practically everywhere throughout astropy.

**Comment by Shaheer-Ahmd** [comment]
@nstarman How would you recommend breaking down the changes? Since many subpackages have this violation, breaking by subpackage would result in a large number of PRs which we don't want aswell.
As Derek mentioned, almost half of the violations are in the `setup_package` file. I can make the changes in these files only throughout the codebase but, still the other half of the violations are mostly in random files (some test configs and other files).

**Comment by eerovaher** [comment]
If a pull request addresses some Ruff rule that has a small number of violations in a large number of sub-packages then it might be best if the general package maintainers (@pllim, @dhomeier and @olebole) perform the final review if not all sub-package maintainers do it in a reasonable amount of time.

**Review by eerovaher — COMMENTED** [review]
This Ruff rule improves consistency, which is helpful both for reading and for grepping. I am in favor of merging this, on the condition that the commits get squashed.

**Comment by pllim** [comment]
The lines touch is small but that is a lot of subpackages. I am worried about this accidentally overwriting some existing variables (`mpl`, `pd`, `np`) though CI passes, so maybe it is ok. Since when ruff gets so nitpicking about import aliases?

I won't have time to review this week myself.

**Comment by pllim** [comment]
I am running cron jobs too, just in case. 🤞

**Comment by nstarman** [comment]
@Shaheer-Ahmd, breaking by subpackage would result in a large number of PRs, yes, but that's not necessarily a bad thing.
But if the general package maintainers are happy to review this, then I'm happy to get this in with 1 PR!

**Review by nstarman — APPROVED** [review]
(Reviewing for units). Now that we have `__all__` defined in the units modules we don't need to worry about `*` imports and thus about making `numpy` a private import.
Therefore this LGTM! Thanks @Shaheer-Ahmd

**Comment by dhomeier** [comment]
> The lines touch is small but that is a lot of subpackages. I am worried about this accidentally overwriting some existing variables (`mpl`, `pd`, `np`) though CI passes, so maybe it is ok. Since when ruff gets so nitpicking about import aliases?

I think the [standard aliases](https://docs.astral.sh/ruff/settings#flake8-import-conventions-aliases) have been around for some time, just not enabled. Seems that overwriting them would raise an [F811](https://docs.astral.sh/ruff/rules/redefined-while-unused/) error or similar, so the replacements here should be safe.
However it will probably not check if variables like `pd` are used in other source files that do not import pandas, which would cause troubles if that import is added at any time later (though with a reasonably clear path to fix them).

Looking at all the "conventional aliases" above, I wonder if we really want to consider them all – `np`, `mpl` and `plt` are definitely standard enough here, `pd` perhaps is already a bit borderline, as it is already a fringe application for many users. But the alias is probably well-known enough; less sure if we always want to worry about keeping `alt` or `px` reserved names (and not reserved for what most astronomers will probably associate them with ;-).
But that should not be a blocker for the rule in general, as the aliases could always be customised in `[tool.ruff.lint.flake8-import-conventions.aliases]`.

For this PR I am only wondering if there was a special reason not the use the `np` alias in any of the `setup_package.py`, and to import it as "private" alias `units`. But unless anyone participating in #370 – @eteq, @perrygreenfield, @astrofrog – recalls the specifics, we probably don't need to stick to that convention.

**Comment by dhomeier** [comment]
> But that should not be a blocker for the rule in general, as the aliases could always be customised in `[tool.ruff.lint.flake8-import-conventions.aliases]`.

Wondering if it would make sense to set our own selection of aliases right away in `.ruff.toml` so we always have an overview of the recommended styles. This could and probably should then be updated in
https://github.com/astropy/astropy/blob/ac222343ee7685a223b74fed7a6e121e0634586a/docs/development/codeguide.rst?plain=1#L201-L207

**Comment by dhomeier** [comment]
> (Reviewing for units). Now that we have `__all__` defined in the units modules we don't need to worry about `*` imports and thus about making `numpy` a private import.

Is that so?
```python
>>> np = 120
>>> stderr = 0.44
>>> from astropy.units.astrophys import *
>>> import math
>>> print(stderr / math.sqrt(np))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: must be real number, not module
```
Not saying this is either likely or recommendable at all, but it doesn't seem to be entirely foolproof.

**Comment by nstarman** [comment]
Not sure how your system is set up that way but that shouldn't be happening.

E.g. in this PR
CleanShot 2024-01-17 at 15 44 23@2x

From Python docs https://www.python.org/doc/essays/packages/

CleanShot 2024-01-17 at 15 49 18@2x

File: astropy/visualization/mpl_normalize.py
Comment (mhvk) [inline astropy/visualization/mpl_normalize.py:29]: To be honest, I have no idea why this line is here! But it makes no sense to assign a different name if we don't even use the import at all.

File: astropy/visualization/wcsaxes/utils.py
Comment (mhvk) [inline astropy/visualization/wcsaxes/utils.py:17]: For a case like this, we don't actually have to import `matplotlib` at all, just do `minversion("matplotlib", "3.8.dev")`
",

File: astropy/wcs/setup_package.py
Comment (mhvk) [inline astropy/wcs/setup_package.py:12]: Doesn't really improve clarity here....

File: astropy/stats/setup_package.py
Comment (mhvk) [inline astropy/stats/setup_package.py:5]: I guess this doesn't strictly hurt, but numpy is really not used to access anything from it here, but rather as the package, to get its include files. By using the full name, it is immediately clear that this is not the usual situation, so this change makes the code less clear, not more.

Below I suggest `from numpy import get_include as get_numpy_include`, but I feel like, again, we are coding around style enforcers, which seems incredibly silly.

File: astropy/io/misc/pandas/connect.py
Comment (mhvk) [inline astropy/io/misc/pandas/connect.py:56]: Not sure this change makes much sense - we're not really using `pandas` here.

File: astropy/io/ascii/core.py
Comment (mhvk) [inline astropy/io/ascii/core.py:26]: These ones are all great.

File: astropy/utils/tests/test_introspection.py
Comment (mhvk) [inline astropy/utils/tests/test_introspection.py:74]: This is just less clear! This is about the package, not anything in it. Would suggest to `#noqa` the import here.

**Review by mhvk — CHANGES_REQUESTED** [review]
In code proper, the changes are good, but a bit less so if the only thing being done is to test whether a package is present. I think we should exclude `setup_package.py` from this style check, or use my suggestion of `from numpy import get_include as get_numpy_include`

p.s. I've never seen the abbreviation `mpl` for matplotlib before, I guess I only really do `import matplotlib.pyplot as plt`...

**Comment by dhomeier** [comment]
Yes, I was aware of that. But had not rebased on #15862 yet. Nevermind

File: astropy/visualization/mpl_normalize.py
Comment (dhomeier) [reply astropy/visualization/mpl_normalize.py:29]: Might be accidentally left over from #3413 – has anyone ever tried to test without this?

**Comment by dhomeier** [comment]
> In code proper, the changes are good, but a bit less so if the only thing being done is to test whether a package is present. I think we should exclude `setup_package.py` from this style check,

Thanks for clarification of the logic behind that! I suggest thus to modify `.ruff.toml` to perhaps
```python
[lint.flake8-import-conventions.aliases]
# Declare the default aliases for ICN001.
matplotlib = "mpl"
"matplotlib.pyplot" = "plt"
numpy = "np"
"numpy.testing" = "npt"
pandas = "pd"

[lint.extend-per-file-ignores]
"__init__.py" = ["E402", "F401", "F403"]
"setup_package.py" = ["ICN001"]  # import-conventions
"test_*.py" = [
```
and then mark the remaining exceptions `#noqa`.
The first aliases are the 3 already recommended by the developer docs; `npt` has been discussed in the context of [helper functions](https://github.com/astropy/astropy/pull/15839#pullrequestreview-1819445324), but I am not really sold on that, as it's only ever been used in a single place so far. Likewise `pd` is a fringe case, but for those who do use pandas, it is absolutely standard.

**Comment by mhvk** [comment]
I agree not to have an abbreviation for `numpy.testing` - we always just import things from it, not use it as a module itself. Also, how about just removing the `matplotlib` abbreviation (keeping the `plt` one)? Or am I the only one who has to do a double-take to think what that means?

File: astropy/visualization/mpl_normalize.py
Comment (dhomeier) [inline astropy/visualization/mpl_normalize.py:29]:
```suggestion
```
Just tested this with and without matplotlib installed, and it seems this can indeed go for good.

**Comment by dhomeier** [comment]
I've been using `mpl` very regularly whenever working on anything outside of the `pyplot` interface, and since it's listed under our [naming conventions](https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions), I'd assume I am not the only one.
Whether it improves clarity in places like `run_astropy_tests.py` is another question.

File: astropy/table/tests/test_table.py
Comment (dhomeier) [inline astropy/table/tests/test_table.py:2156]: Would it make sense to use `minversion("pandas", "2.0")` from `astropy.utils.introspection` here as well to remove the `import pandas` altogether?

File: astropy/stats/setup_package.py
Comment (eerovaher) [reply astropy/stats/setup_package.py:5]: There's no need to clutter `pyproject.toml` with this.

File: astropy/table/tests/test_table.py
Comment (eerovaher) [reply astropy/table/tests/test_table.py:2156]: Yes, then we wouldn't need to import `Version` either.

File: astropy/utils/tests/test_introspection.py
Comment (eerovaher) [reply astropy/utils/tests/test_introspection.py:74]: In practice `numpy` is imported as `np` much more often than not, so it's not bad to test that `minversion(np, version)` works.

File: astropy/wcs/setup_package.py
Comment (eerovaher) [reply astropy/wcs/setup_package.py:12]: I don't see what's unclear about `import numpy as np`.

File: astropy/stats/setup_package.py
Comment (dhomeier) [reply astropy/stats/setup_package.py:5]: No, we are making all customisations in `.ruff.toml`, right?

Comment (eerovaher) [reply astropy/stats/setup_package.py:5]: `.ruff.toml` is meant to be a temporary file and the long-term goal is to get rid of it entirely. All the permanent configuration is in `pyproject.toml`.

File: astropy/utils/tests/test_introspection.py
Comment (dhomeier) [reply astropy/utils/tests/test_introspection.py:74]: If it would not point to the same package and version under the alias, something would be very fishy!
But just below there is in fact the explicit test for `minversion("numpy", version)` to complement it, so leaving `np` in place here might make sense.

Comment (eerovaher) [reply astropy/utils/tests/test_introspection.py:74]: What I meant to say was that `minversion(np, ...)` might be closer to real-life usage than `minversion(numpy, ...)`.

File: astropy/stats/setup_package.py
Comment (dhomeier) [reply astropy/stats/setup_package.py:5]: Hmm, there are already some 70 per-file and per-subpackage exclusions in there – are those all meant to be temporary? 😳

File: astropy/wcs/setup_package.py
Comment (mcara) [reply astropy/wcs/setup_package.py:12]: _"Doesn't really improve clarity here"_ does not imply (at least to me) that `import numpy as np` is "unclear". But in this particular case, I do find `as np` an unnecessary alias.

File: astropy/stats/setup_package.py
Comment (eerovaher) [reply astropy/stats/setup_package.py:5]: For the most part yes. It is possible that we might want to ignore a small number of rules currently listed in `.ruff.toml` permanently, but whenever we make such a decision the relevant ignores should be moved to `pyproject.toml`. If there is some rule we usually want to enforce then it is best to use the `noqa` instructions for the exceptions.

File: astropy/utils/tests/test_introspection.py
Comment (dhomeier) [reply astropy/utils/tests/test_introspection.py:74]: Well, this test is actually not about how one commonly imports numpy, but testing the functionality of `minversion` itself with its different interfaces. In that regard it might even make sense to keep both imports and also test `minversion(numpy, version)` along with `minversion(np, version)`, but perhaps that's overkill...

File: astropy/wcs/setup_package.py
Comment (dhomeier) [reply astropy/wcs/setup_package.py:12]: The point is, I think, that even if it is not (much) unclearer, there was a conscious decision to choose this style even when the `np` alias was already standard usage, possibly to emphasise the correspondence to `include_dirs=["numpy"]` (#9726).

File: astropy/stats/setup_package.py
Comment (dhomeier) [reply astropy/stats/setup_package.py:5]: I'd certainly find it preferable then to "clutter" `pyproject.toml` with one exception line than to add `noqa`s to every single `setup_package.py`.

Comment (eerovaher) [reply astropy/stats/setup_package.py:5]: Most importantly [Ruff rule RUF100](https://docs.astral.sh/ruff/rules/unused-noqa/) checks for unnecessary `noqa` instructions, so if anything ever changes in Ruff itself or in our configuration then our code will not be littered with obsolete `noqa` comments. But `pyproject.toml` will have to be edited manually.

Comment (mhvk) [reply astropy/stats/setup_package.py:5]: Agreed with @dhomeier -  I prefer not to be distracted by unnecessary `noqa`. So, we
1. Work around this rule with `from numpy import get_include as get_numpy_include`;
2. Ignore `setup_package.py`; or
3. Just make the few good changes here and simply keep ignoring `ICN001` - this is *not* a big deal one, and new contributions are very unlikely to "break" it.

Comment (eerovaher) [reply astropy/stats/setup_package.py:5]: I'd be fine with `from numpy import get_include as get_numpy_include` or `from numpy import get_include as numpy_get_include`.

File: astropy/utils/tests/test_introspection.py
Comment (mhvk) [reply astropy/utils/tests/test_introspection.py:74]: OK, for this case, happy to just go with `np` - @eerovaher is right that this will be the more common usage.

File: astropy/wcs/setup_package.py
Comment (Shaheer-Ahmd) [reply astropy/wcs/setup_package.py]: > Agreed with @dhomeier - I prefer not to be distracted by unnecessary `noqa`. So, we
> 
> 1. Work around this rule with `from numpy import get_include as get_numpy_include`;
> 2. Ignore `setup_package.py`; or
> 3. Just make the few good changes here and simply keep ignoring `ICN001` - this is _not_ a big deal one, and new contributions are very unlikely to "break" it.


We have changed this to `from numpy import get_include as get_numpy_include` since the package is not being used itself, only its include files are needed.

File: astropy/stats/setup_package.py
Comment (dhomeier) [reply astropy/stats/setup_package.py]: I second your initial comment that creating some awkward custom imports merely to keep a linting tool happy or at least quiet feels rather silly; therefore I am not comfortable with the `get_numpy_include` option 1.
No. 2 would be my preferred solution – `pyproject.toml` already has 90+ lines of Ruff configuration with 20-something per-file ignores, so I don't see how another ignore could be considered unacceptable clutter. Same for 6 or 7 lines to clearly declare our own endorse aliases.
I probably could even live with `noqa` as next best option, rather than keeping the rule ignored (which would also require a new entry in `pyproject.toml`). We may not have needed it in any relevant code, and it is indeed unlikely a PR breaking it would be accepted, but it is just the purpose of an automated checker like this to ensure they really don't without anyone needing to think about it any more.

**Comment by Shaheer-Ahmd** [comment]
So, what would be the better choice? 
Use `get_numpy_include/numpy_get_include` 
or ignore ICN001 in `setup_package.py` files altogether.

**Comment by dhomeier** [comment]
Ah, seeing you already beat me in implementing the alternative!
But my preference is also to add a per-file-ignore for `"setup_package.py"`.
Should wait for the other reviewers to add their opinions.

**Comment by eerovaher** [comment]
We shouldn't write weird code just to satisfy a linter, but there is nothing strange about `import numpy as np` and there is nothing strange about `from {module} import {function}` either.

**Comment by dhomeier** [comment]
No, but would anyone have chosen to alias it to something like `get_numpy_include` in normal circumstances?

**Comment by eerovaher** [comment]
We could just use `from numpy import get_include`.

**Comment by mhvk** [comment]
I'm fine with the current state, somewhat preferring the renaming (all this may become irrelevant anyway, as surely at some point someone is going to suggest we switch to `meson`...)

File: astropy/io/misc/pandas/connect.py
Comment (mhvk) [reply astropy/io/misc/pandas/connect.py:56]: This one is not important enough, really, let's just leave it as you have it.

**Review by mhvk — APPROVED** [review]
I think this is now all fine! Thanks, @Shaheer-Ahmd 

p.s. Very small suggestion for future PRs: push commits in batches so one avoids running CI for every commit separately.

File: astropy/units/astrophys.py
Comment (pllim) [inline astropy/units/astrophys.py:17]: @mhvk , are you sure the numpy import wasn't down here for a specific reason?

Comment (dhomeier) [reply astropy/units/astrophys.py:17]: If it was to keep numpy private here that would be resolved by https://github.com/astropy/astropy/pull/15899#pullrequestreview-1827893978.

File: astropy/visualization/wcsaxes/utils.py
Comment (pllim) [inline astropy/visualization/wcsaxes/utils.py:1]: This change broke pyinstaller. Please revert.

https://github.com/astropy/astropy/actions/runs/7584151828/job/20657236613?pr=15899

File: astropy/units/astrophys.py
Comment (mhvk) [reply astropy/units/astrophys.py:17]: It was originally there to ensure it did not pollute the namespace. But with #15862, we define an `__all__`, so we can do whatever we want. Indeed, arguably we should have changed this in #15862!

File: astropy/visualization/wcsaxes/utils.py
Comment (mhvk) [reply astropy/visualization/wcsaxes/utils.py:1]: So weird, that seemed like there might be a bug in `minversion`, but I cannot reproduce the failure, so hard to investigate. Anyway, for this PR indeed best to just revert.

Comment (dhomeier) [reply astropy/visualization/wcsaxes/utils.py:1]: Yes, obviously matplotlib is installed and `module = resolve_name('matplotlib')` has succeeded at this point, yet `metadata.Distribution.from_name('matplotlib')` is raising `importlib.metadata.PackageNotFoundError` – though that may not necessarily indicate an `ImportError`
> :raises PackageNotFoundError: When the named package's distribution
            metadata cannot be found.

Wondering if this has something to do with mpl's rather nonstandard `matplotlib._version` interface, but cannot reproduce this either even with my old system Python 3.9.6...

**Comment by Shaheer-Ahmd** [comment]
Is this merge-ready then? Any other changes?

**Comment by dhomeier** [comment]
If everyone agrees on the custom alias, that should be ready, but needs squeezing. If you want to do it yourself, reducing to 2-3 commits should probably be ok.
And if we stick with the full set of enforced aliases rather than setting our own, they should be added to the documentation (see https://github.com/astropy/astropy/pull/15899#issuecomment-1896540747).
Or perhaps add a link there to the Ruff docs, as I suspect this may become a moving target.
The failed cron job is _probably_ unrelated, but not certain...

**Comment by Shaheer-Ahmd** [comment]
By the "full set of enforced aliases" you mean
`{"altair": "alt", "matplotlib": "mpl", "matplotlib.pyplot": "plt", "numpy": "np", "pandas": "pd", "seaborn": "sns", "tensorflow": "tf", "tkinter": "tk", "holoviews": "hv", "panel": "pn", "plotly.express": "px", "polars": "pl", "pyarrow": "pa"}`
these? (default aliases in ruff docs)
and for the squeezing, should I make a new PR with larger but less in number commits and link it here? (sorry for asking too many questions, I am a newbie here)

**Comment by dhomeier** [comment]
> By the "full set of enforced aliases" you mean `{"altair": "alt", "matplotlib": "mpl", "matplotlib.pyplot": "plt", "numpy": "np", "pandas": "pd", "seaborn": "sns", "tensorflow": "tf", "tkinter": "tk", "holoviews": "hv", "panel": "pn", "plotly.express": "px", "polars": "pl", "pyarrow": "pa"}` these? (default aliases in ruff docs)

That is the full list, but since most of these are not imported anywhere within astropy, and many likely never will be, it should be enough to add `pandas as pd` to our list, and then link to https://docs.astral.sh/ruff/settings/#flake8-import-conventions-aliases for reference – that way we don't need to worry about updating whenever Ruff developers discover a new alias considered conventional.

> and for the squeezing, should I make a new PR with larger but less in number commits and link it here? (sorry for asking too many questions, I am a newbie here)

No, squashing (sorry, my typo before) is done on your existing branch, as described in
https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#how-to-squash
One commit for `units`, one for `setup_package.py` and one for everything else would probably make sense, though I am not sure if you can combine them out of order...
We can also squash automatically on merge, but that way everything would be stuffed into a single commit.

**Comment by dhomeier** [comment]
Thanks a lot, looking great now! Apparently there is a [new deprecation in numpy-dev](https://github.com/astropy/astropy/actions/runs/7604280412/job/20706989683?pr=15899#step:10:103), but that's nothing that needs to concern this PR.

**Review by dhomeier — APPROVED** [review]

File: docs/development/codeguide.rst
Comment (eerovaher) [inline docs/development/codeguide.rst:206]: If anybody ever deviates from the naming conventions ICN001 enforces then the Ruff hook in `pre-commit` will clearly state where in the code the violation occurred and what name it expects to be used. It is even possible to have Ruff fix the violations automatically. I don't think this text in the developer documentation is informative anymore.

Comment (dhomeier) [reply docs/development/codeguide.rst:206]: Personally I prefer to know what I am expected to write before a pre-commit tool tells me I've done it wrong and how to fix it. Besides, we are still seeing issues with contributors who are unable to set up their `pre-commit` locally.
If we have section in the docs stating in some detail what import styles Ruff is or is going to enforce via `isort`, `flake8` conventions and what else, that might replace this explicit list, but currently I don't see such a reference. Opinions may vary on this.

Comment (eerovaher) [reply docs/development/codeguide.rst:206]: I'm not sure if people are actually having problems installing `pre-commit`. I consider it much more likely that they haven't followed our instructions to install `pre-commit` simply because they have not read our developer documentation at all, in which case nothing we write there is going to help them.

I'd rather not write down any specific rules that Ruff would be enforcing anyways because I believe the shorter our developer documentation is the more people might actually read it.

Comment (nstarman) [reply docs/development/codeguide.rst:206]: I suspect similarly. I've installed pre-commit on everything from a Python env on my phone (bad idea, don't do dev work on the phone) to a compute cluster. I've never once run into a problem with the pip install except when I didn't have git on a machine.

Comment (dhomeier) [reply docs/development/codeguide.rst:206]: Currently the docs do not even state that you must install and activate pre-commit before you are allowed to push anything upstream, so that would need to be changed along the way. And then how would you handle cases when a series of commits fail in the CI pre-commit, you ask the contributor to run pre-commit locally and they state they have it installed, yet it is apparently allowing invalid commits? I thought the point of all the pre-commit tools was to lower the threshold for new contributors, but that is not always reflected by my experience – apparently it is not that easy for everyone, or are those just some freak cases of exotic Windows installations?

Comment (eerovaher) [reply docs/development/codeguide.rst:206]: > Currently the docs do not even state that you must install and activate pre-commit before you are allowed to push anything upstream, so that would need to be changed along the way.

In our documentation we indeed only say that we _encourage_ installing `pre-commit`, and I do think we should change that.

> And then how would you handle cases when a series of commits fail in the CI pre-commit, you ask the contributor to run pre-commit locally and they state they have it installed, yet it is apparently allowing invalid commits?

In such a case we can conclude that they have not installed our `pre-commit` hooks. In my usage `pre-commit` has proven itself to be trustworthy. A first-time contributor who needs multiple commits to get the code style checks to pass has proven themselves to be untrustworthy when it comes to code style and the related tooling.

> I thought the point of all the pre-commit tools was to lower the threshold for new contributors, but that is not always reflected by my experience...

`pre-commit` can only help those who use it.

> ... apparently it is not that easy for everyone, or are those just some freak cases of exotic Windows installations?

@pllim, do you know how simple or complicated it is to install `pre-commit` on Windows?

Updating our developer documentation properly is beyond the scope of this pull request, so I agree with merging this as is, assuming that the commits get squashed.

Comment (pllim) [reply docs/development/codeguide.rst:206]: Re: Windows -- I think I used both conda and pip on different times and they both worked, but I mostly work on WSL2.

Comment (pllim) [reply docs/development/codeguide.rst:206]: Is this conversation blocking merge? Should I wait?

Comment (nstarman) [reply docs/development/codeguide.rst:206]: >  A first-time contributor who needs multiple commits to get the code style checks to pass has proven themselves to be untrustworthy when it comes to code style and the related tooling.

It is technically possible to get multiple commits to pass and then fail pre-commit. But this requires rebases and other fancy stuff that then skip the "pre-commit".
Our CI run of pre-commit should tell any contributor exactly what to fix. And the comment `pre-commit autofix` can take care of many stuff server side.
Installing pre-commit locally AFAIK is always as easy as `pip install pre-commit; cd <git repo>; pre-commit install` and makes their life much easier.

Comment (eerovaher) [reply docs/development/codeguide.rst:206]: > Is this conversation blocking merge? Should I wait?

I think we should update some of our developer documentation, but doing that properly is beyond the scope here and shouldn't block merging.

**Comment by pllim** [comment]
Thanks, all!

File: docs/development/codeguide.rst
Comment (dhomeier) [reply docs/development/codeguide.rst:206]: Indeed, this is moving to an entirely different topic and certainly needs more discussion.
The PR was already squashed; I probably would not have added the `as np` imports just to rename them later, but I think our commit history can live with that. ;-)

**Comment by dhomeier** [comment]
Hmm, maybe I misunderstood the interpretation of the "rebase or squash necessary?" checkmark...

**Comment by pllim** [comment]
@dhomeier , what do you mean?

**Comment by dhomeier** [comment]
I normally check the marks in the sense of "no further action required" since @Shaheer-Ahmd had already squashed down to 3 commits locally. But that list is only used very inconsistently anyway.

**Comment by pllim** [comment]
Oh... I only noticed that the box in the origin post was not check. 😬 

`By checking this box, the PR author has requested that maintainers do NOT use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.`

**Comment by dhomeier** [comment]
But that box does not mean "if unchecked the PR will always be squashed on merge"? We surely have smaller PRs merged with more commits...

**Comment by pllim** [comment]
Sorry, it is my bad. I didn't know you wanted it not squashed. I made a call at merge time and apparently I made the wrong call. I will take full responsibility.

But seems overkill to revert and I don't want to force push to main. I will be more careful next time!

**Comment by dhomeier** [comment]
No problem, I think it can fit into a single commit – I meant to reply to Eero that it was already squashed, but did not get to it while the discussion on pre-commit was still going on. Thanks everyone!

File: docs/development/codeguide.rst
Comment (dhomeier) [reply docs/development/codeguide.rst:206]: To conclude this part of the discussion here, in my experience it simply is not straightforward for every first-time contributor to parse a report like https://results.pre-commit.ci/run/github/2081289/1703509121.Z6WkqI4KQVm9d2te2s3qhg and make the required fixes, and if I see such commits coming in even after mentioning the use of `pre-commit`, I am still hesitant to label anyone as "untrustworthy" without knowing myself about any possible pitfalls on a given platform. For which we don't seem to be able to provide that much guidance e.g. for the various Windows platforms.

**Pull request merged**
